File: src/tools/pmd-cpd/eval-repos/synthetic/go/semantic_dup_literals.go
// Package synthetic contains Go files for CPD testing with semantic duplicates (literals).
package synthetic

import (
	"math"
	"regexp"
	"strings"
)

// CalculateBronzeTierDiscount calculates discount for bronze tier customers.
func CalculateBronzeTierDiscount(price float64) float64 {
	baseDiscount := 5.0
	maxDiscount := 15.0
	threshold := 100.0

	if price < threshold {
		return price * (1 - baseDiscount/100)
	}

	additional := (price - threshold) * 0.02
	totalDiscount := math.Min(baseDiscount+additional, maxDiscount)
	return price * (1 - totalDiscount/100)
}

// CalculateSilverTierDiscount calculates discount for silver tier - semantic duplicate with different literals.
func CalculateSilverTierDiscount(price float64) float64 {
	baseDiscount := 10.0
	maxDiscount := 25.0
	threshold := 150.0

	if price < threshold {
		return price * (1 - baseDiscount/100)
	}

	additional := (price - threshold) * 0.02
	totalDiscount := math.Min(baseDiscount+additional, maxDiscount)
	return price * (1 - totalDiscount/100)
}

// CalculateGoldTierDiscount calculates discount for gold tier - semantic duplicate with different literals.
func CalculateGoldTierDiscount(price float64) float64 {
	baseDiscount := 15.0
	maxDiscount := 35.0
	threshold := 200.0

	if price < threshold {
		return price * (1 - baseDiscount/100)
	}

	additional := (price - threshold) * 0.02
	totalDiscount := math.Min(baseDiscount+additional, maxDiscount)
	return price * (1 - totalDiscount/100)
}

// Address represents an address.
type Address struct {
	Street     string
	City       string
	State      string
	Zip        string
	Province   string
	PostalCode string
}

// ValidateUSAddress validates US address format.
func ValidateUSAddress(address Address) []string {
	var errors []string
	requiredFields := []string{"street", "city", "state", "zip"}
	statePattern := regexp.MustCompile(`^[A-Z]{2}$`)
	zipPattern := regexp.MustCompile(`^\d{5}(-\d{4})?$`)

	for _, field := range requiredFields {
		var value string
		switch field {
		case "street":
			value = address.Street
		case "city":
			value = address.City
		case "state":
			value = address.State
		case "zip":
			value = address.Zip
		}
		if value == "" {
			errors = append(errors, strings.Title(field)+" is required")
		}
	}

	if address.State != "" && !statePattern.MatchString(address.State) {
		errors = append(errors, "State must be 2 letter code")
	}
	if address.Zip != "" && !zipPattern.MatchString(address.Zip) {
		errors = append(errors, "ZIP must be 5 digits")
	}

	return errors
}

// ValidateCAAddress validates Canadian address - semantic duplicate with different literals.
func ValidateCAAddress(address Address) []string {
	var errors []string
	requiredFields := []string{"street", "city", "province", "postalCode"}
	provincePattern := regexp.MustCompile(`^[A-Z]{2}$`)
	postalPattern := regexp.MustCompile(`^[A-Z]\d[A-Z] ?\d[A-Z]\d$`)

	for _, field := range requiredFields {
		var value string
		switch field {
		case "street":
			value = address.Street
		case "city":
			value = address.City
		case "province":
			value = address.Province
		case "postalCode":
			value = address.PostalCode
		}
		if value == "" {
			errors = append(errors, strings.Title(field)+" is required")
		}
	}

	if address.Province != "" && !provincePattern.MatchString(address.Province) {
		errors = append(errors, "Province must be 2 letter code")
	}
	if address.PostalCode != "" && !postalPattern.MatchString(address.PostalCode) {
		errors = append(errors, "Postal code must be A1A 1A1 format")
	}

	return errors
}

File: src/tools/pmd-cpd/eval-repos/synthetic/rust/semantic_dup_literals.rs
//! Rust file with semantic duplicates - same logic with different literal values.

use regex::Regex;

/// Calculate discount for bronze tier customers.
pub fn calculate_bronze_tier_discount(price: f64) -> f64 {
    let base_discount = 5.0;
    let max_discount = 15.0;
    let threshold = 100.0;

    if price < threshold {
        return price * (1.0 - base_discount / 100.0);
    }

    let additional = (price - threshold) * 0.02;
    let total_discount = (base_discount + additional).min(max_discount);
    price * (1.0 - total_discount / 100.0)
}

/// Calculate discount for silver tier - semantic duplicate with different literals.
pub fn calculate_silver_tier_discount(price: f64) -> f64 {
    let base_discount = 10.0;
    let max_discount = 25.0;
    let threshold = 150.0;

    if price < threshold {
        return price * (1.0 - base_discount / 100.0);
    }

    let additional = (price - threshold) * 0.02;
    let total_discount = (base_discount + additional).min(max_discount);
    price * (1.0 - total_discount / 100.0)
}

/// Calculate discount for gold tier - semantic duplicate with different literals.
pub fn calculate_gold_tier_discount(price: f64) -> f64 {
    let base_discount = 15.0;
    let max_discount = 35.0;
    let threshold = 200.0;

    if price < threshold {
        return price * (1.0 - base_discount / 100.0);
    }

    let additional = (price - threshold) * 0.02;
    let total_discount = (base_discount + additional).min(max_discount);
    price * (1.0 - total_discount / 100.0)
}

/// Address representation.
pub struct Address {
    pub street: String,
    pub city: String,
    pub state: String,
    pub zip: String,
    pub province: String,
    pub postal_code: String,
}

/// Validate US address format.
pub fn validate_us_address(address: &Address) -> Vec<String> {
    let mut errors = Vec::new();
    let required_fields = ["street", "city", "state", "zip"];
    let state_pattern = Regex::new(r"^[A-Z]{2}$").unwrap();
    let zip_pattern = Regex::new(r"^\d{5}(-\d{4})?$").unwrap();

    for field in &required_fields {
        let value = match *field {
            "street" => &address.street,
            "city" => &address.city,
            "state" => &address.state,
            "zip" => &address.zip,
            _ => continue,
        };
        if value.is_empty() {
            let capitalized = field.chars().next().unwrap().to_uppercase().to_string() + &field[1..];
            errors.push(format!("{} is required", capitalized));
        }
    }

    if !address.state.is_empty() && !state_pattern.is_match(&address.state) {
        errors.push("State must be 2 letter code".to_string());
    }
    if !address.zip.is_empty() && !zip_pattern.is_match(&address.zip) {
        errors.push("ZIP must be 5 digits".to_string());
    }

    errors
}

/// Validate Canadian address - semantic duplicate with different literals.
pub fn validate_ca_address(address: &Address) -> Vec<String> {
    let mut errors = Vec::new();
    let required_fields = ["street", "city", "province", "postal_code"];
    let province_pattern = Regex::new(r"^[A-Z]{2}$").unwrap();
    let postal_pattern = Regex::new(r"^[A-Z]\d[A-Z] ?\d[A-Z]\d$").unwrap();

    for field in &required_fields {
        let value = match *field {
            "street" => &address.street,
            "city" => &address.city,
            "province" => &address.province,
            "postal_code" => &address.postal_code,
            _ => continue,
        };
        if value.is_empty() {
            let capitalized = field.chars().next().unwrap().to_uppercase().to_string() + &field[1..];
            errors.push(format!("{} is required", capitalized));
        }
    }

    if !address.province.is_empty() && !province_pattern.is_match(&address.province) {
        errors.push("Province must be 2 letter code".to_string());
    }
    if !address.postal_code.is_empty() && !postal_pattern.is_match(&address.postal_code) {
        errors.push("Postal code must be A1A 1A1 format".to_string());
    }

    errors
}
